lib/SILOptimizer/FunctionSignatureTransforms/ExistentialTransform.cpp

```diff
@@ -0,0 +1,595 @@
+    }
+    InterfaceErrorResult =
+        SILResultInfo(MappedTy, FTy->getErrorResult().getConvention());
+  }
+
+  /// Now add the yields.
+  llvm::SmallVector<SILYieldInfo, 8> InterfaceYields;
+  for (SILYieldInfo InterfaceYield : FTy->getYields()) {
+    InterfaceYields.push_back(InterfaceYield);
+  }
+
+  /// Finally the ExtInfo.
+  auto ExtInfo = FTy->getExtInfo();
+  ExtInfo = ExtInfo.withRepresentation(SILFunctionTypeRepresentation::Thin);
+  auto witnessMethodConformance = FTy->getWitnessMethodConformanceOrNone();
+
+  /// Return the new signature.
+  return SILFunctionType::get(
+      NewGenericSig, ExtInfo, FTy->getCoroutineKind(),
+      FTy->getCalleeConvention(), InterfaceParams, InterfaceYields,
+      InterfaceResults, InterfaceErrorResult, Ctx, witnessMethodConformance);
+}
+
+/// Create the Thunk Body with always_inline attribute.
+void ExistentialTransform::populateThunkBody() {
+
+  SILModule &M = F->getModule();
+
+  F->setThunk(IsSignatureOptimizedThunk);
+  F->setInlineStrategy(AlwaysInline);
+
+  /// Remove original body of F.
+  for (auto It = F->begin(), End = F->end(); It != End;) {
+    auto *BB = &*It++;
+    removeDeadBlock(BB);
+  }
+
+  /// Create a basic block and the function arguments.
+  auto *ThunkBody = F->createBasicBlock();
+  for (auto &ArgDesc : ArgumentDescList) {
+    ThunkBody->createFunctionArgument(ArgDesc.Arg->getType(), ArgDesc.Decl);
+  }
+
+  /// Builder to add new instructions in the Thunk.
+  SILBuilder Builder(ThunkBody);
+  SILOpenedArchetypesTracker OpenedArchetypesTracker(F);
+  Builder.setOpenedArchetypesTracker(&OpenedArchetypesTracker);
+  Builder.setCurrentDebugScope(ThunkBody->getParent()->getDebugScope());
+
+  /// Location to insert new instructions.
+  auto Loc = ThunkBody->getParent()->getLocation();
+
+  /// Create the function_ref instruction to the NewF.
+  auto *FRI = Builder.createFunctionRef(Loc, NewF);
+
+  auto GenCalleeType = NewF->getLoweredFunctionType();
+  auto CalleeGenericSig = GenCalleeType->getGenericSignature();
+  auto OrigGenCalleeType = F->getLoweredFunctionType();
+  auto OrigCalleeGenericSig = OrigGenCalleeType->getGenericSignature();
+
+  /// Determine arguments to Apply.
+  /// Generate opened existentials for generics.
+  llvm::SmallVector<SILValue, 8> ApplyArgs;
+  llvm::SmallDenseMap<GenericTypeParamType *, Type> Generic2OpenedTypeMap;
+  for (auto &ArgDesc : ArgumentDescList) {
+    auto iter = Arg2GenericTypeMap.find(ArgDesc.Index);
+    auto it = ExistentialArgDescriptor.find(ArgDesc.Index);
+    if (iter != Arg2GenericTypeMap.end() &&
+        it != ExistentialArgDescriptor.end()) {
+      ArchetypeType *Opened;
+      auto OrigOperand = ThunkBody->getArgument(ArgDesc.Index);
+      auto SwiftType = ArgDesc.Arg->getType().getASTType();
+      auto OpenedType =
+          SwiftType->openAnyExistentialType(Opened)->getCanonicalType();
+      auto OpenedSILType = NewF->getModule().Types.getLoweredType(OpenedType);
```

Comment:
Maybe just F.getLoweredType(OpenedType)

Reply:
This does not work!

Reply:
Why not?

Reply:
There is no API in F that allows you to take a Type. Its an empty one.

Reply:
That is F.getLoweredType does not take an argument.